io/io/src/TFileMerger.cxx

```diff
@@ -543,9 +555,10 @@ Bool_t TFileMerger::MergeOne(TDirectory *target, TList *sourcelist, Int_t type,
    Bool_t canBeMerged = kTRUE;
 
    TList dirtodelete;
-   auto getDirectory = [&dirtodelete](TDirectory *parent, const char *name, const TString &pathname)
+   auto getDirectory = [&dirtodelete](TDirectory *parent, const TString &pathname)
    {
-      TDirectory *result = dynamic_cast<TDirectory*>(parent->GetList()->FindObject(name));
+      auto pathPieces = ROOT::Split(pathname, "\\/");
+      TDirectory *result = LookupObjByFullPath(parent, pathname, pathPieces);
```

Comment:
Is there a concrete difference between this and 'just' executing directly line 550/563:
```
          result = parent->GetDirectory(pathname);
```
?

The original idea of the call to `FindObject` was to save time if the answer was findable without parsing (which seems reintroduced by the split)

Reply:
Actually no, I was remembering wrong ... the intent is completely different and is broken by the new code.

The core of the lambda function is to determine whether the `TDirectory` we are looking for is already in memory or needs to be loaded from file and hence added to the list of things to cleanup.

The `parent->GetList()->FindObject(name)` is the part that checks whether it is already in memory while the call to `GetDirectory` will return the requested pointer either from memory or after loading it from disk.

See commit e97dc3678ae9da6628242afceca3142d6c319832 and the (incomplete first part) 30fd4c79425c8be12be7af6fe1936317d1f5eec7.

This might be fixable by tweaks the new line 444.

Reply:
I don't understand how calling `FindObject` can work if it only considers the basename of the directory. Besides, even with my change I'm still going through the logic at lines 563-567, so the adding to the delete list is still happening, no?

Reply:
Actually, thinking about it, couldn't `getDirectory()` be just this?
```
   auto getDirectory = [&dirtodelete](TDirectory *parent, const TString &pathname)
   {
      TDirectory *result = parent->GetDirectory(pathname);
      if (result && result != parent)
         dirtodelete.Add(result);
      return result;
   };

```

Reply:
> The FindObject part is wrong because it will incorrectly return an existing directory Foo even when we're looking for Bar/Baz/Foo, 

This is indeed the bug we are trying to fix here.

> but it looks to me like we can just omit it..?

Not exactly.  Calling `FindObject` before/instead of `GetDirectory` is a way to cheaply get the answer to the question `did GetDirectory allocate a TDirectory or return an existing one?`.

> Unless the idea is that we mustn't add the directory to the list if it was already in memory; but in this case can't we simply deduplicate the elements from dirtodelete?

Yes this will likely have the right semantic.  However we have to make this work efficiently with file with upwards of `O(100,000)` directories in them ...